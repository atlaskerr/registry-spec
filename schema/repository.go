package schema

import (
	"github.com/xeipuuv/gojsonschema"
)

func repository() *gojsonschema.Schema {
	path := "repository/repository.schema.json"
	mediaType := MediaTypeRepository
	return load(path, mediaType)
}

func repositoryCreate() *gojsonschema.Schema {
	path := "repository/repository.create.schema.json"
	mediaType := MediaTypeRepositoryCreate
	return load(path, mediaType)
}

func repositoryList() *gojsonschema.Schema {
	path := "repository/repository.list.schema.json"
	mediaType := MediaTypeRepositoryList
	return load(path, mediaType)
}
