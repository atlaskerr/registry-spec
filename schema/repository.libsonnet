// Copyright Â© 2019 Titan Distribution Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

local types = import 'types.libsonnet';

local openapi = 'openapi';
local jsonschema = 'jsonschema';
local V7 = 'http://json-schema.org/draft-07/schema#';

local jid = {
  repo: 'https://openpackages.io/schema/repo',
  createRepo: 'https://openpackages.io/schema/repo-create',
  listRepos: 'https://openpackages.io/schema/repo-list',
};


local repositorySchemaFunc(output=jsonschema) = {
  [if output == jsonschema then '$id']: jid.repo,
  [if output == jsonschema then '$schema']: V7,
  type: 'object',
  additionalProperties: false,
  properties: {
    name: { type: 'string' },
    namespace: { type: 'string' },
    project: { type: 'string' },
    size: types.uint64,
    package: {
      type: 'object',
      additionalProperties: false,
      properties: {
        type: { type: 'string' },
        icon: { type: 'string' },
        documentation: { type: 'string' },
      },
    },
    labels: types.mapStringString(output),
  },
};

local repositoryExample = {
  name: 'repository-example-foo',
  namespace: 'namespace-example-foo',
  project: 'project-example-foo',
  package: {
    type: 'OCI Image',
    icon: 'https://opencontainers/static/icon-small.png',
    documentation: 'https://github.com/opencontainers/image-spec',
  },

  size: 34252334,
  labels: {
    team: 'team-example',
    maintainer: 'Billy Seiken',
    costCenter: 'cs-foo',
  },
};

local repositoryCreateSchemaFunc(output=jsonschema) = {
  [if output == jsonschema then '$id']: jid.createRepo,
  [if output == jsonschema then '$schema']: V7,
  type: 'object',
  additionalProperties: false,
  properties: {
    type: { type: 'string' },
    labels: types.labels(output),
  },
};

local repositoryCreateExample = {
  type: 'OCI Image',
  labels: {
    provider: {
      awsAccount: 'aws-account-foo',
      pricingPlan: 'PREMIUM',
      region: 'us-east-1',
    },
    consumer: {
      team: 'team-foo',
      manager: 'Tom Ripen',
      costCenter: 'cs-foo',
    },
  },
};

local repositoryListSchemaFunc(output=jsonschema) = {
  [if output == jsonschema then '$id']: jid.listRepos,
  [if output == jsonschema then '$schema']: V7,
  type: 'object',
  additionalProperties: false,
  properties: {
    repositories: {
      type: 'array',
      items: repositorySchemaFunc(''),
    },
  },
};

local repositoryListExample = {
  repositories: [
    {
      name: 'app-foo',
      namespace: 'namespace-foo',
      project: 'project-foo',
      package: {
        type: 'oci-image-v1.0.0',
        icon: 'https://opencontainers/static/icon-small.png',
        documentation: 'https://github.com/opencontainers/image-spec',
      },
      labels: {
        provider: {
          awsAccount: 'aws-account-foo',
          pricingPlan: 'PREMIUM',
          region: 'us-east-1',
        },
        consumer: {
          team: 'team-foo',
          manager: 'Tom Ripen',
          costCenter: 'cs-foo',
        },
      },
    },
    {
      name: 'app-bar',
      namespace: 'namespace-foo',
      project: 'project-foo',
      package: {
        type: 'docker-image-v2.2.0',
        icon: 'https://docker.io/static/icon-small.png',
        documentation: 'https://github.com/moby/moby/blob/master/image/spec/v1.2.md',
      },
      labels: {
        provider: {
          awsAccount: 'aws-account-foo',
          pricingPlan: 'PREMIUM',
          region: 'us-east-1',
        },
        consumer: {
          team: 'team-bar',
          manager: 'Tom Ripen',
          costCenter: 'cs-foo',
        },
      },
    },
    {
      name: 'app-foo-helm',
      namespace: 'namespace-foo',
      project: 'project-foo',
      package: {
        type: 'helm-chart-v1.0.0',
        icon: 'https://helm.sh/static/icon-small.png',
        documentation: 'https://helm.sh/docs',
      },
      labels: {
        provider: {
          awsAccount: 'aws-account-foo',
          pricingPlan: 'PREMIUM',
          region: 'us-east-1',
        },
        consumer: {
          team: 'team-helm',
          manager: 'Tom Ripen',
          costCenter: 'cs-foo',
        },
      },
    },
  ],
};

{
  repository:: {
    schema:: repositorySchemaFunc,
    example:: repositoryExample,
  },
  repositoryCreate:: {
    schema:: repositoryCreateSchemaFunc,
    example:: repositoryCreateExample,
  },
  repositoryList:: {
    schema:: repositoryListSchemaFunc,
    example:: repositoryListExample,
  },
}
