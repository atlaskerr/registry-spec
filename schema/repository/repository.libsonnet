// Copyright Â© 2019 Titan Distribution Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

local types = import '../types.libsonnet';

local openapi = 'openapi';
local jsonschema = 'jsonschema';
local V7 = 'http://json-schema.org/draft-07/schema#';

local jid = {
  repo: 'https://openpackages.io/schema/repo',
  createRepo: 'https://openpackages.io/schema/repo-create',
  listRepos: 'https://openpackages.io/schema/repo-list',
};

local repo(output=jsonschema) = {
  [if output == jsonschema then '$id']: jid.repo,
  [if output == jsonschema then '$schema']: V7,
  type: 'object',
  properties: {
    name: { type: 'string' },
    namespace: { type: 'string' },
    project: { type: 'string' },
    labels: types.mapStringString(output),
  },
};

local createRepo(output=jsonschema) = {
  [if output == jsonschema then '$id']: jid.createRepo,
  [if output == jsonschema then '$schema']: V7,
  type: 'object',
  properties: {
    type: { type: 'string' },
    labels: types.mapStringString(output),
  },
};

local listRepos(output=jsonschema) = {
  [if output == jsonschema then '$id']: jid.listRepos,
  [if output == jsonschema then '$schema']: V7,
  type: 'object',
  properties: {
    repositories: {
      type: 'array',
      items: repo(''),
    },
  },
};

{
  repo:: repo,
  createRepo:: createRepo,
  listRepos:: listRepos,
}
