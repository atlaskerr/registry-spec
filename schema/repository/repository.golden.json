{
  "name": "repository-example-foo",
  "namespace": "namespace-example-foo",
  "project": "project-example-foo",
  "size": 34252334,
  "package": {
    "type": "OCI Image",
    "documentation": "https://github.com/opencontainers/image-spec",
    "icon": "https://opencontainers/static/icon-small.png",
    "clients": {
      "docker": {
        "command": "docker",
        "actions": {
          "download": [
            "pull",
            "{package}:{tag}"
          ],
          "upload": [
            "push",
            "{package}:{tag}"
          ]
        }
      },
      "exampleCli": {
        "command": "example-cli",
        "actions": {
          "download": [
            "download",
            "--image",
            "{package}",
            "--tag",
            "{tag}"
          ],
          "upload": [
            "upload",
            "--image",
            "{package}",
            "--tag",
            "{tag}"
          ]
        }
      },
      "skopeo": {
        "command": "skopeo",
        "actions": {
          "download": [
            "pull",
            "{package}:{tag}"
          ],
          "upload": [
            "push",
            "{package}:{tag}"
          ]
        }
      }
    }
  },
  "labels": {
    "costCenter": "cs-foo",
    "maintainer": "Billy Seiken",
    "team": "team-example"
  }
}
